pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+	return err
+}
+
+func (c *Controller) deleteGauges(labelSlice []prometheus.Labels) {
+	for _, labels := range labelSlice {
+		allocatableGaugeVec.Delete(labels)
+		requestsGaugeVec.Delete(labels)
+		limitsGaugeVec.Delete(labels)
+		overheadGaugeVec.Delete(labels)
+	}
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node, resourceTypeName string) (prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+	metricLabels[resourceType] = resourceTypeName
+	metricLabels[nodeName] = node.GetName()
+	metricLabels[nodeProvisioner] = node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	metricLabels[nodeZone] = node.Labels[v1.LabelTopologyZone]
+	metricLabels[nodeArchitecture] = node.Labels[v1.LabelArchStable]
+	metricLabels[nodeCapacityType] = node.Labels[v1alpha5.LabelCapacityType]
+	metricLabels[nodeInstanceType] = node.Labels[v1.LabelInstanceTypeStable]
+	metricLabels[nodePhase] = string(node.Status.Phase)
+	// Add node labels
+	labels, err := json.Marshal(node.GetLabels())
+	if err != nil {
+		return nil, fmt.Errorf("marshal pod labels: %w", err)
+	}
+	metricLabels[nodeLabels] = string(labels)
+	return metricLabels, nil
+}
+
+func (c *Controller) updateGauges(ctx context.Context, node *v1.Node) error {
+
+	// Calculate total requests and limits from all pods
+	podlist := &v1.PodList{}
+	if err := c.KubeClient.List(ctx, podlist, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
+		return fmt.Errorf("listing pods on node %s, %w", node.Name, err)
+	}
+	reqs, limits := getPodsTotalRequestsAndLimits(podlist.Items)
+	allocatable := node.Status.Capacity
+	overheads := calculateDaemonSetOverhead(podlist.Items)
+	if len(node.Status.Allocatable) > 0 {
+		allocatable = node.Status.Allocatable
+		// calculating system daemons overhead
+		for resourceName, quantity := range allocatable {
+			overhead := node.Status.Capacity[resourceName]
+			overhead.Sub(quantity)
+			overheads[resourceName] = overhead
+		}
+	}
+
+	// Populate overhead metrics
+	if err := c.insertGaugeValues(overheads, node, overheadGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate total request metrics
+	if err := c.insertGaugeValues(reqs, node, requestsGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate total limits metrics
+	if err := c.insertGaugeValues(limits, node, limitsGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+	// Populate allocatable metrics
+	if err := c.insertGaugeValues(allocatable, node, allocatableGaugeVec); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to generate gauge: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Controller) insertGaugeValues(resources map[v1.ResourceName]resource.Quantity, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+
+	for resourceName, quantity := range resources {
```

Comment:
Minor style nit, but I typically find that functions whose entire body is a for loop should probably instead have the for loop outside. It forces all of the logic to be indented, which hampers readability. It also tends to be simpler to reason about shortcircuiting, etc.

Reply:
I agree with you. But here is my counter argument for this case.
Resources in K8S libraries are mainly described using v1.ResourceList. Most functions return this type. I think it will be more consistent to also use v1.ResourceList as input and output variables.

Also taking the for loop outside will creates quite a few very repetitive for loops that looks like what I have written in CS101.

Reply:
Totally optional. see my other comment though:
```
for vec, resources := range map[*prometheus.GaugeVec]map[v1.ResourceName]resource.Quantity{
   overheadGaugeVec: overhead,
   requestsGaugeVec: requests,
   ...
}{
  for ... {
      c.insert...
  }
}
```

Reply:
Great suggestion! Done.